libraries/AP_AIS/AP_AIS.cpp

```diff
@@ -0,0 +1,764 @@
+        return;
+    }
+
+    _uart->begin(AP::serialmanager().find_baudrate(AP_SerialManager::SerialProtocol_AIS, 0));
+}
+
+// update AIS, expected to be called at 20hz
+void AP_AIS::update()
+{
+    if (!_uart || !enabled()) {
+        return;
+    }
+
+    // read any available lines
+    u_int32_t nbytes = MAX(_uart->available(),1024U);
+    while (nbytes-- > 0) {
+        const int16_t byte = _uart->read();
+        if (byte == -1) {
+            break;
+        }
+        const char c = byte;
+        if (decode(c)) {
+            const bool log_all = (_options & AIS_OPTIONS_LOG_ALL_RAW) != 0;
+            const bool log_unsupported = ((_options & AIS_OPTIONS_LOG_UNSUPPORTED_RAW) != 0) && !log_all; // only log unsupported if not logging all
+
+            if (_incoming.total  > AIVDM_BUFFER_SIZE)  {
+                // no point in trying to decode it wont fit
+                if (log_all || log_unsupported) {
+                    log_raw(&_incoming);
+                }
+                break;
+            }
+            if (log_all) {
+                log_raw(&_incoming);
+            }
+
+            if (_incoming.num == 1 && _incoming.total == 1) {
+                // single part message
+                if (!payload_decode(_incoming.payload) && log_unsupported) {
+                    // could not decode so log
+                    log_raw(&_incoming);
+                }
+            } else if (_incoming.num == _incoming.total) {
+                // last part of a multi part message
+                uint8_t index = 0;
+                uint8_t msg_parts[_incoming.total - 1];
+                for  (uint8_t i = 0; i < AIVDM_BUFFER_SIZE; i++) {
+                    // look for the rest of the message from the start of the buffer
+                    // we assume the mesage has be received in the correct order
+                    if (_AIVDM_buffer[i].num == (index + 1) && _AIVDM_buffer[i].total == _incoming.total && _AIVDM_buffer[i].ID == _incoming.ID) {
+                        msg_parts[index] = i;
+                        index++;
+                        if (index >= _incoming.num) {
+                            break;
+                        }
+                    }
+                }
+
+                // did we find the right number?
+                if (_incoming.num != index) {
+                    // could not find all of the message, save messages
+                    if (log_unsupported) {
+                        for (uint8_t i = 0; i < index; i++) {
+                            log_raw(&_AIVDM_buffer[msg_parts[i]]);
+                        }
+                        log_raw(&_incoming);
+                    }
+                    // remove
+                    for (uint8_t i = 0; i < index; i++) {
+                        buffer_shift(msg_parts[i]);
+                    }
+                    break;
+                }
+
+                // combine packets
+                char multi[AIVDM_PAYLOAD_SIZE*_incoming.total];
```

Comment:
This makes no sense to me.  Elsewhere in the code we're comparing `_coming.total` against `AIVDM_PAYLOAD_SIZE` and here we're multiplying the two together?!

Also, unless I've missed a sanity check this can allocate `65*65` == 4225 bytes on the stack based on user input.  That sounds very bad.

Reply:
_incoming.total is the number off ADVIM messages the AIS packet it spread over. We need to combine the payload of all the packets to decode (or at least that is how I did it here).   _incoming.total must be less than AIVDM_BUFFER_SIZE, I have that set to 10. So its only 650 bytes in the worst case.